draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -356,869 +524,832 @@ document are to be interpreted as described in {{RFC2119}}.
+7. If gx2 is square, set x = x2 and y = sqrt(gx2)
+8. Output (x, y)
 ~~~
 
-**Implementation**
+#### Implementation
 
 The following procedure implements the Simple SWU's algorithm in a straight-line
 fashion.
 
 ~~~
 map2curve_simple_swu(alpha)
+Input: alpha, an octet string to be hashed.
+Output: (x, y), a point on E.
 
-Input:
-
-  alpha - value to be encoded, an octet string
-
-Output:
-
-  (x, y) - a point in E
-
-Precomputations:
-
-1.  c1 = -B / A mod p           // Field arithmetic
-2.  c2 = (p - 1)/2              // Integer arithmetic
+Constants:
+1.  c1 = - B / A
 
 Steps:
-
-1.    u = hash2base(alpha)  // {0,1}^* -> Fp
-2.   u2 = u^2
-3.   u2 = -u2                // u2 = -u^2
-4.   u4 = u2^2
-5.   t1 = u4 + u2
-6.   t1 = t1^(-1)
-7.   n1 = 1 + t2             // n1 = 1 + (1 / (u^4 - u^2))
-8.   x1 = c1 * n1            // x1 = -B/A * (1 + (1 / (u^4 - u^2)))
-9.  gx1 = x1 ^ 3
-10.  t1 = A * x1
-11. gx1 = gx1 + t1
-12. gx1 = gx1 + B            // gx1 = x1^3 + Ax1 + B = g(x1)
-13.   x2 = u2 * x1           // x2 = -u^2 * x1
-14.  gx2 = x2^3
-15.   t1 = A * x2
-16.  gx2 = gx2 + 12
-17.  gx2 = gx2 + B           // gx2 = x2^3 + Ax2 + B = g(x2)
-18.   e = gx1^c2
-19.   x = CMOV(x1, x2, l1)      // If l1 = 1, choose x1, else choose x2
-20.  gx = CMOV(gx1, gx2, l1)    // If l1 = 1, choose gx1, else choose gx2
-21.   y = sqrt(gx)
-22. Output (x, y)
+1.    u = hash2base(alpha)
+2.   t1 = -u^2            
+3.   t2 = t1^2            
+4.   x1 = t2 + t1
+5.   x1 = 1 / x1
+6.   x1 = x1 + 1         
+7.   x1 = x1 * c1            // x1 = (-B / A) * (1 + (1 / (u^4 - u^2)))
+8.  gx1 = x1^2
+9.  gx1 = gx1 + A
+10. gx1 = gx1 * x1
+11. gx1 = gx1 + B            // gx1 = x1^3 + A*x1 + B
+12.  x2 = t1 * x1            // x2 = -u^2 * x1
+13. gx2 = x2^2
+14. gx2 = gx2 + A
+15. gx2 = gx2 * x2
+16. gx2 = gx2 + B            // gx2 = x2^3 + A*x2 + B
+17.   e = is_square(gx1, q)
+18.   x = CMOV(x2, x1, e)    // If e=True, x = x1, else x = x2
+19.  gx = CMOV(gx2, gx1, e)  // If e=True, gx = gx1, else gx = gx2
+20.   y = sqrt(gx, q)
+21. Output (x, y)
 ~~~
 
-### Boneh-Franklin Method {#supersingular}
-
-The map2curve_bf(alpha) implements the Boneh-Franklin method {{BF01}} which
-covers the case of supersingular curves `E: y^2=x^3+B`. This method does not
-guarantee that the resulting a point be in a specific subgroup of the curve.
-To do that, a scalar multiplication by a cofactor is required.
-
-**Preconditions**
-
-This algorithm works for any Weierstrass curve over `F_q` such that `A=0` and
-`q=2 mod 3`.
-
-**Examples**
+## Encodings for Montgomery curves
 
-- `y^2 = x^3 + 1`
+### Elligator2 Method {#elligator2}
 
-**Algorithm**: map2curve_bf
+The map2curve_elligator2(alpha) implements the Elligator2 {{Elligator2}} for
+curves defined by y^2 = x^3 + A\*x^2 + B\*x such that A\*B\*(A^2 - 4\*B) != 0.
+In particular, this method applies to the Montgomery curves y^2 = x^3 + A\*x^2 + x
+setting B=1.
 
-Input:
+Preconditions: A Montgomery curve  such that A!=0.
```

Comment:
nit: extra space between `curve` and `space`

Reply:
done. However parser eats extra spaces.